examples/federated-learning/players.py

```diff
@@ -0,0 +1,155 @@
+import functools
+
+import tensorflow as tf
+import tf_encrypted as tfe
+
+from util import UndefinedModelFnError
+
+
+class Owner(type):
+
+  def __call__(self, *args, **kwargs):
+    owner_obj = type.__call__(*args, **kwargs)
+
+    # Decorate user-defined TF function that needs to be pinned to a particular
+    # device and compiled
+    if hasattr(owner_obj, "aggregator_fn"):
+      self.handle_tf_local_fn(owner_obj, "aggregator_fn")
+
+    if hasattr(owner_obj, "evaluator_fn"):
+      self.handle_tf_local_fn(owner_obj, "evaluator_fn")
+
+    # Decorate user-defined TFE local_computations
+    if has_attr(obj, "model_fn"):
+      self.handle_tfe_local_fn(owner_obj, "model_fn")
```

Comment:
this smells bad:
- behavior being redefined in base class, away from the user
- base class referring to methods only defined in subclasses

i suggest getting rid of `Owner` as follows:
- turn `pin_to_owner` into a global decorator
- introduce decorated `_call_agggregator_fn`, `_call_evaluator_fn`, and `_call_model_fn`
- use the above in `_runner` and `_update_one_round`

Reply:
Yeah, that seems right! The benefit of using a metaclass is that it would enable this behavior in any class created from itself, but it's not much of a benefit here since we know the only two player types we want.

Reply:
But I'd say that in general the two points you mentioned are not necessarily bad when emphasized to the user -- tf.Variable does the first, and torch.nn.Module does both. Both of these expect to be subclassed in a variety of different ways, and so behavior is modified so as to make those subclasses comply with extra behavior inside of contexts and the like.